### PR TITLE
WIP: Add ProgressPreference to SpecialVariables var and type arrays

### LINKED ISSUE #20657: The new `-ProgressAction` common parameter breaks calls to advanced scripts and functions.
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

`-ProgressAction Ignore` should translate into a target scope-local `$ProgressPreference = 'Ignore'` variable, analogously to how it works with other common parameter, such as `-ErrorAction`.

```powershell
& { [CmdletBinding()]param() $ProgressPreference } -ProgressAction Ignore
```

### Expected behavior

```console
Ignore
```

### Actual behavior

The following error occurs:

```none
: Specified argument was out of the range of valid values. (Parameter 'index')
```

### Error details

_No response_

### Environment data

```powershell
PowerShell 7.4.1
```

### Visuals

_No response_

### LINKED ISSUE #21074: The -ProgressAction parameter doesn't work (PowerShell 7.4.1 only)
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Upon using the -ProgressAction switch, PowerShell returns an error message regardless of correct formatting. Please see the examples below.

### Expected behavior

```
PS > Expand-Archive -Path .\1.zip -DestinationPath ".\1" -ProgressAction 'SilentlyContinue'

PS > Repair-Volume -DriveLetter c -Scan -ProgressAction 'SilentlyContinue'

NoErrorsFound
```

### Actual behavior

```
PS > Expand-Archive -Path .\1.zip -DestinationPath ".\1" -ProgressAction 'SilentlyContinue'

Expand-Archive: Cannot convert value "SilentlyContinue" to type "System.Collections.IEnumerator". Error: "Invalid cast from 'System.Management.Automation.ActionPreference' to 'System.Collections.IEnumerator'."

PS > Repair-Volume -DriveLetter c -Scan -ProgressAction 'SilentlyContinue'

Repair-Volume: Cannot convert value "SilentlyContinue" to type "System.Collections.IEnumerator". Error: "Invalid cast from 'System.Management.Automation.ActionPreference' to 'System.Collections.IEnumerator'."
```

### Error details

```
Exception             :
    Type           : System.Management.Automation.PSInvalidCastException
    ErrorRecord    :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Cannot convert value "SilentlyContinue" to type "System.Collections.IEnumerator". Error: "Invalid cast from 'System.Management.Automation.ActionPreference'
to 'System.Collections.IEnumerator'."
            HResult : -2146233087
        CategoryInfo          : InvalidArgument: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : InvalidCastIConvertible
    TargetSite     :
        Name          : ConvertIConvertible
        DeclaringType : System.Management.Automation.LanguagePrimitives
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message        : Cannot convert value "SilentlyContinue" to type "System.Collections.IEnumerator". Error: "Invalid cast from 'System.Management.Automation.ActionPreference' to
'System.Collections.IEnumerator'."
    InnerException :
        Type       : System.InvalidCastException
        TargetSite :
            Name          : DefaultToType
            DeclaringType : System.Convert
            MemberType    : Method
            Module        : System.Private.CoreLib.dll
        Message    : Invalid cast from 'System.Management.Automation.ActionPreference' to 'System.Collections.IEnumerator'.
        Source     : System.Private.CoreLib
        HResult    : -2147467262
        StackTrace :
   at System.Convert.DefaultToType(IConvertible value, Type targetType, IFormatProvider provider)
   at System.Management.Automation.LanguagePrimitives.ConvertIConvertible(Object valueToConvert, Type resultType, Boolean recursion, PSObject originalValueToConvert,
IFormatProvider formatProvider, TypeTable backupTable)
    Source         : System.Management.Automation
    HResult        : -2147467262
    StackTrace     :
   at System.Management.Automation.LanguagePrimitives.ConvertIConvertible(Object valueToConvert, Type resultType, Boolean recursion, PSObject originalValueToConvert,
IFormatProvider formatProvider, TypeTable backupTable)
   at System.Management.Automation.LanguagePrimitives.ConversionData`1.Invoke(Object valueToConvert, Type resultType, Boolean recurse, PSObject originalValueToConvert,
IFormatProvider formatProvider, TypeTable backupTable)
   at System.Management.Automation.LanguagePrimitives.ConvertTo(Object valueToConvert, Type resultType, Boolean recursion, IFormatProvider formatProvider, TypeTable
backupTypeTable)
   at System.Management.Automation.LanguagePrimitives.ConvertTo[T](Object valueToConvert)
   at System.Management.Automation.MutableTuple`64.SetValueImpl(Int32 index, Object value)
   at System.Management.Automation.PSScriptCmdlet.SetPreferenceVariables()
   at System.Management.Automation.PSScriptCmdlet.BeginProcessing()
   at System.Management.Automation.Cmdlet.DoBeginProcessing()
   at System.Management.Automation.CommandProcessorBase.DoBegin()
CategoryInfo          : InvalidArgument: (:) [Repair-Volume], PSInvalidCastException
FullyQualifiedErrorId : InvalidCastIConvertible,Repair-Volume
InvocationInfo        :
    MyCommand        : Repair-Volume
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : 2
    Line             : Repair-Volume -DriveLetter c -Scan -ProgressAction 'SilentlyContinue'
    Statement        : Repair-Volume -DriveLetter c -Scan -ProgressAction 'SilentlyContinue'
    PositionMessage  : At line:1 char:1
                       + Repair-Volume -DriveLetter c -Scan -ProgressAction 'SilentlyContinue'
                       + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    InvocationName   : Repair-Volume
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

### Environment data

```
Name                           Value
----                           -----
PSVersion                      7.4.1
PSEdition                      Core
GitCommitId                    7.4.1
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

Output

### PR DESCRIPTION
**Pull request by cmkb3**
# PR Summary

This fixes an array out of bounds/type error that occurs when using the -ProgressAction common parameter in advanced functions. The reference variable and variable type arrays are missing the ProgressPreference entry which causes various engine errors including index out of bounds and invalid typecast from index collision.

## PR Context

Fixes: #21074 'The -ProgressAction parameter doesn't work (PowerShell 7.4.1 only)'
Fixes: #20657 'The new -ProgressAction common parameter breaks calls to advanced scripts and functions. '

## PR Checklist

- [✅] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [✅] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [✅] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [✅] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): 
- **User-facing changes**
  - [✅] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [✅] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: 
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: 
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: 
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed:

### CODE REVIEW

**Comment by cmkb3** [comment]
> @cmkb3 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

File: test/powershell/Language/Scripting/CommonParameters.Tests.ps1 [hunk 1]
```diff
@@ -147,6 +147,40 @@ Describe "Common parameters support for script cmdlets" -Tags "CI" {
         }
     }
 
+    Context "ProgressAction" {
+        It "Silences with no parameters/variables" {
```
Comment (iSazonov) [inline test/powershell/Language/Scripting/CommonParameters.Tests.ps1:151]: Please reword tests names. "Silence" is not explain that test checks.

**Review by iSazonov — COMMENTED** [review]
Thanks for finding and patching the issue.

I suggest to discuss replacing `Progress = 16,` with `Progress = 8,` in `internal enum PreferenceVariable` for performance reason (with the PR the tuple size jumps from 16 to 32).

**Comment by cmkb3** [comment]
> Thanks for finding and patching the issue.
> 
> I suggest to discuss replacing `Progress = 16,` with `Progress = 8,` in `internal enum PreferenceVariable` for performance reason (with the PR the tuple size jumps from 16 to 32).

If I understand correctly, the Enum value must be unique within the combination of both `AutomaticVariable` and `PreferenceVariable` internal enum sets. Value 8 already exists within `AutomaticVariable` as `PSCommandPath`.

    internal enum AutomaticVariable
     {
        Underbar = 0,
        Args = 1,
        This = 2,
        Input = 3,
        PSCmdlet = 4,
        PSBoundParameters = 5,
        MyInvocation = 6,
        PSScriptRoot = 7,
        PSCommandPath = 8,
        NumberOfAutomaticVariables // 1 + the last, used to initialize global scope.
    }

    internal enum PreferenceVariable
    {
        Debug = 9,
        Verbose = 10,
        Error = 11,
        WhatIf = 12,
        Warning = 13,
        Information = 14,
        Confirm = 15,
        Progress = 16,
    }

**Comment by iSazonov** [comment]
> If I understand correctly,

Yes, I see. Could you please add new Assert like in line 718 for `PreferenceVariableTypes`?

If we can not use value less 16 we will allocate double size tuples so we have to think about whether we need to fix the performance issue and how.

**Comment by jborean93** [comment]
> If we can not use value less 16 we will allocate double size tuples so we have to think about whether we need to fix the performance issue and how.

While we should think about this the damage is already done by accepting and implementing the new parameter in 7.4 . We should fix this issue first then investigate the performance side after not hold off a bugfix for a problem today.

**Comment by cmkb3** [comment]
> Yes, I see. Could you please add new Assert like in line 718 for `PreferenceVariableTypes`?

What is this in reference to? Neither of the changed files are above 500 lines; searching for 'Assert' in either returned no results for me.

**Comment by iSazonov** [comment]
> > Yes, I see. Could you please add new Assert like in line 718 for `PreferenceVariableTypes`?
> 
> What is this in reference to? Neither of the changed files are above 500 lines; searching for 'Assert' in either returned no results for me.

Ah, sorry! https://github.com/PowerShell/PowerShell/blob/c2f1ff00ae43bc81b4e26168792a1e1118c0d4ed/src/System.Management.Automation/engine/parser/VariableAnalysis.cs#L718-L721


> > If we can not use value less 16 we will allocate double size tuples so we have to think about whether we need to fix the performance issue and how.
> 
> While we should think about this the damage is already done by accepting and implementing the new parameter in 7.4 . We should fix this issue first then investigate the performance side after not hold off a bugfix for a problem today.

Currently we haven't the performance issue because the feature doesn't work. We should either revert https://github.com/PowerShell/PowerShell/pull/18887 or fix without any (perf) regression.

**Comment by jborean93** [comment]
> Currently we haven't the performance issue because the feature doesn't work. We should either revert https://github.com/PowerShell/PowerShell/pull/18887 or fix without any (perf) regression.

It works just fine for binary cmdlets, it’s advanced functions where it fails and this fixes. We only loose out on the size 16 tuple if the function didn’t have any paramaters or variables (or was un-optimisable). I would say the vast majority of advanced functions will have at least one parameter or one variable making it go beyond 16 in any case.

**Comment by cmkb3** [comment]
> > > Yes, I see. Could you please add new Assert like in line 718 for `PreferenceVariableTypes`?
> > 
> > 
> > What is this in reference to? Neither of the changed files are above 500 lines; searching for 'Assert' in either returned no results for me.
> 
> Ah, sorry!
> 
> https://github.com/PowerShell/PowerShell/blob/c2f1ff00ae43bc81b4e26168792a1e1118c0d4ed/src/System.Management.Automation/engine/parser/VariableAnalysis.cs#L718-L721
> 
> > > If we can not use value less 16 we will allocate double size tuples so we have to think about whether we need to fix the performance issue and how.
> > 
> > 
> > While we should think about this the damage is already done by accepting and implementing the new parameter in 7.4 . We should fix this issue first then investigate the performance side after not hold off a bugfix for a problem today.
> 
> Currently we haven't the performance issue because the feature doesn't work. We should either revert #18887 or fix without any (perf) regression.

Are you looking for something like this? Asserts that the enums match the associated array names/types lengths.

```cs
Diagnostics.Assert(AutomaticVariable.NumberOfAutomaticVariables == SpecialVariables.AutomaticVariables.Length
    && AutomaticVariable.NumberOfAutomaticVariables == SpecialVariables.AutomaticVariableTypes.Length,
    "automatic variable enum length does not match both automatic variable and automatic variable types length.");

Diagnostics.Assert(SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariables.Length
    && SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariableTypes.Length,
    "preference variable enum length does not match both preference variable and preference variable types length.").
```

**Comment by iSazonov** [comment]
> Are you looking for something like this (@iSazonov)? Asserts that the enums match the associated array names/types lengths.

Yes. We should add check for types you change in the PR.

Have you checked the existing assert in DEBUG mode? I guess it will throw now.

**Comment by jborean93** [comment]
> Have you checked the existing assert in DEBUG mode? I guess it will throw now.

The debug statements

```
Diagnostics.Assert(AutomaticVariable.NumberOfAutomaticVariables == SpecialVariables.AutomaticVariables.Length
    && AutomaticVariable.NumberOfAutomaticVariables == SpecialVariables.AutomaticVariableTypes.Length,
    "automatic variable enum length does not match both automatic variable and automatic variable types length.");

Diagnostics.Assert(SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariables.Length
    && SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariableTypes.Length,
    "preference variable enum length does not match both preference variable and preference variable types length.").
```

Do not throw as they are just checking that the slots used for the automatic vars and action preferences match up together. In this case the enum wasn't populated with the progress preference var so the slots still match up with what was in the enum.

If we wanted to add a debug assertion we could potentially do add something to the below which somehow checks the enums defined [here](https://github.com/PowerShell/PowerShell/blob/c2f1ff00ae43bc81b4e26168792a1e1118c0d4ed/src/System.Management.Automation/engine/SpecialVariables.cs#L413-L423). I don't know off a good check for this off the top of my head though.

https://github.com/PowerShell/PowerShell/blob/c2f1ff00ae43bc81b4e26168792a1e1118c0d4ed/src/System.Management.Automation/engine/parser/VariableAnalysis.cs#L154-L159

**Comment by cmkb3** [comment]
> The debug statements
> 
> [...]
>
> Do not throw as they are just checking that the slots used for the automatic vars and action preferences match up together. In this case the enum wasn't populated with the progress preference var so the slots still match up with what was in the enum.

If the second debug statement I proposed existed in the current codebase (pre-merge), it should have thrown, if I understand correctly.

```cs
// Enum, length of 8
internal enum PreferenceVariable
{
    Debug = 9,
    Verbose = 10,
    Error = 11,
    WhatIf = 12,
    Warning = 13,
    Information = 14,
    Confirm = 15,
    Progress = 16,
};

// Array of preference values, length of 7
internal static readonly string[] PreferenceVariables =
{
    SpecialVariables.DebugPreference,
    SpecialVariables.VerbosePreference,
    SpecialVariables.ErrorActionPreference,
    SpecialVariables.WhatIfPreference,
    SpecialVariables.WarningPreference,
    SpecialVariables.InformationPreference,
    SpecialVariables.ConfirmPreference,
};

// Array of preference variable types, length of 7
internal static readonly string[] PreferenceVariables =
{
    SpecialVariables.DebugPreference,
    SpecialVariables.VerbosePreference,
    SpecialVariables.ErrorActionPreference,
    SpecialVariables.WhatIfPreference,
    SpecialVariables.WarningPreference,
    SpecialVariables.InformationPreference,
    SpecialVariables.ConfirmPreference,
};

// Debug assert, effectively becomes "Assert 8 == 7 && 8 == 7"
Diagnostics.Assert(SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariables.Length
    && SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariableTypes.Length,
    "preference variable enum length does not match both preference variable and preference variable types length.").
    ```

**Comment by jborean93** [comment]
Ah my apologies I misread things.

**Comment by iSazonov** [comment]
@cmkb3 I wonder why do you change nuget configs? Do you use `Start-PSBuild -UseNuGetOrg`?

**Comment by cmkb3** [comment]
> @cmkb3 I wonder why do you change nuget configs? Do you use `Start-PSBuild -UseNuGetOrg`?

Yes, I'm a git idiot and accidentally pushed those configs when trying to not do that. Tried to undo it and I've only made it worse so far, I'm afraid 😩

**Comment by iSazonov** [comment]
@cmkb3 You can close the PR and open new one with clean commit.

**Pull request closed without merging**